doc/migration.md

```diff
@@ -0,0 +1,426 @@
+Alternatively you may provide an object (or objects) with `operator()` implementations which handle the event subtypes you want to process.
+Notably you do not have to provide callbacks for all possible event subtypes.
+Depending on the current active event subtype, the corresponding callback is called.
+Here's what that looks like:
+
+```cpp
+const auto onClose = [&window](const sf::Event::Closed&)
+{
+    window.close();
+};
+
+const auto onKeyPressed = [&window](const sf::Event::KeyPresssed& keyPressed)
+{
+    if (keyPressed.scancode == sf::Keyboard::Scancode::Escape)
+        window.close();
+};
+
+while (window.isOpen())
+{
+    window.handleEvents(onClose, onKeyPressed);
+
+    // Remainder of main loop
+}
+```
+
+## Window Styles and States
+
+A new `sf::State` enumeration was added for specifying the state of the window which means whether the window is floating or fullscreen.
+Here's a before-and-after showing how this affects constructing a window.
+
+v2:
+```cpp
+sf::RenderWindow window(sf::VideoMode::getFullscreenModes().at(0), "Title", sf::Style::Fullscreen);
+```
+
+v3:
+```cpp
+sf::RenderWindow window(sf::VideoMode::getFullscreenModes().at(0), "Title", sf::State::Fullscreen);
+```
+
+## Scoped Enumerations
+
+SFML 3 converts all old style unscoped enumerations to scoped enumerations.
+This improves the type safety of the interface.
+This means that the name of the enumeration is now part of the namespace required to access values of that enumeration.
+
+For example, take the enumeration `sf::Keyboard::Key`.
+`sf::Keyboard::A` becomes `sf::Keyboard::Key::A`.
+The name of the enumeration now appears as a nested namespace when accessing one of the enumeration's values.
+
+Here is a complete list of all enumerations which have undergone this change:
+
+* `sf::BlendMode::Equation`
+* `sf::BlendMode::Factor`
+* `sf::Cursor::Type`
+* `sf::Ftp::Response::Status`
+* `sf::Ftp::TransferMode`
+* `sf::Http::Request::Method`
+* `sf::Http::Response::Status`
+* `sf::Joystick::Axis`
+* `sf::Keyboard::Key`
+* `sf::Keyboard::Scan`
+* `sf::Mouse::Button`
+* `sf::Mouse::Wheel`
+* `sf::PrimitiveType`
+* `sf::Sensor::Type`
+* `sf::Shader::Type`
+* `sf::Socket::Status`
+* `sf::Socket::Type`
+* `sf::SoundSource::Status`
+* `sf::VertexBuffer::Usage`
+
+## `sf::Rect<T>`
+
+`sf::Rect<T>` has been refactored from 4 scalar values into 2 `sf::Vector2<T>`s.
+Its two data members are `sf::Vector2<T>`s named `position` and `size`.
+This means that `sf::Rect<T>::getPosition()` and `sf::Rect<T>::getSize()` have been removed in favor of directly accessing the `position` and `size` data members.
+The 4-parameter constructor was also removed in favor of the constructor which takes two `sf::Vector2<T>`s.
+
+v2:
+```cpp
+sf::FloatRect rect(10, 20, 30, 40);
+sf::Vector2f position = rect.getPosition();
+sf::Vector2f size = rect.getSize();
+```
+
+v3:
+```cpp
+sf::FloatRect rect({10, 20}, {30, 40});
+sf::Vector2f [position, size] = rect;
```

Comment:
This not valid C++. You need `auto`.
```suggestion
auto [position, size] = rect;
```

Also, I would have kept the `rect.position` and `rect.size` example lines and added the structured binding as an additional example. Otherwise, it is not clear by skimming through code that you are not forced to use structured bindings to get position or size.

Reply:
> added the structured binding as an additional example

It seems a bid odd to highlight this as something new when it's possible with SFML 2 as well. I think it's out of scope to dedicate space to highlight how `sf::Rect` still supports destructuring.